SConstruct

```diff
@@ -454,7 +464,57 @@ def add_cluster_analysis(w):
     def partis_seqmeta(outdir, c):
         return c['_process_partis'][2]
 
+    # Partis alternative naives
+    # -------------------------
 
+    @w.add_target()
+    def alternative_naive_probabilities(outdir, c):
+        '''
+        Write partis alternative naives to a fasta in order of probability
+        '''
+        if c['cluster']['naive_probabilities']:
+            cluster_name = c['cluster'].get('seed_name', c['cluster']['id'])
+
+            naives_sorted_by_prob =  list(sorted(c['cluster']['naive_probabilities'], key=lambda x: x[1], reverse=True))
+            def write_naive_fastas(target, source, env):
```

Comment:
Fine to use action functions like this, though keep in mind, SCons can't do it's fancy diffing to figure out when targets need to update as easily this way.

Reply:
Yeah I learned a bit about this since this target was getting unconditionally rebuilt until I used `list()` and took `naives_sorted_by_prob` out of the action function. I did that after reading https://scons-users.scons.narkive.com/uoYdUQNw/why-is-this-action-always-run though the fact that I'm not sure if both of those changes were necessary is evidence for me not completely understanding how it decides to rebuild. My understanding was that one should ensure that a callable action function should be as immutable as possible so that SCons can know when it has actually been changed.

Reply:
Interesting... maybe SCons actually _can_ guess when things have changed in some cases. I didn't realize that. It doesn't seem to have to do with immutability as much as details of the function object, and what sort of metadata it has which can be used to determine whether anything looks like it's changed.

One thing you may need to be careful of though, is that if you use a helper function in the action function definition, and then change the helper function, it might not need to know that the target needs to get rebuilt. You might be able to get around this by using `env.Depends`, but I've never tried it with action functions or their helpers like this.

Note that this same problem can still crop up with scripts though, if you have one script that actually requires or depends upon another. Here though, I _do_ know that you can use `env.Depends` to ensure things get rebuilt as needed, but it's easy to forget to do this. So worth keeping in mind.

Reply:
Good to know. Yeah the docs are not very clear on whether this applies to callable action functions:

>env.Depends(target, dependency)
>Specifies an explicit dependency; the target will be rebuilt whenever the dependency has changed. Both the specified target and dependency can be a string (usually the path name of a file or directory) or Node objects, or a list of strings or Node objects (such as returned by a Builder call). This should only be necessary for cases where the dependency is not caught by a Scanner for the file.